daemon/nvidia_linux.go

```diff
@@ -0,0 +1,104 @@
+package daemon
+
+import (
+	"strconv"
+
+	"github.com/containerd/containerd/contrib/nvidia"
+	"github.com/docker/docker/pkg/capabilities"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+)
+
+// TODO: add list of device capabilities in daemon/node info
+
+var nvidiaDriver = &deviceDriver{
+	capset:     capabilities.Set{"gpu": true},
+	updateSpec: setNvidiaGPUs,
+}
+
+var errConflictCountDeviceIDs = errors.New("cannot set both Count and DeviceIDs on device request")
+
+// These are NVIDIA-specific capabilities stolen from github.com/containerd/containerd/contrib/nvidia.capFlags
+var allNvidiaCaps = map[nvidia.Capability]struct{}{
+	nvidia.Compute:  {},
+	nvidia.Compat32: {},
+	nvidia.Graphics: {},
+	nvidia.Utility:  {},
+	nvidia.Video:    {},
+	nvidia.Display:  {},
+}
+
+func init() {
+	for c := range nvidiaDriver.capset {
+		nvidiaDriver.capset[string(c)] = true
+	}
+}
+
+func setNvidiaGPUs(s *specs.Spec, dev *deviceInstance) error {
+	var opts []nvidia.Opts
+
+	req := dev.req
+	if req.Count != "" && len(req.DeviceIDs) > 0 {
+		return errConflictCountDeviceIDs
+	}
+
+	if len(req.DeviceIDs) > 0 {
+		var ids []int
+		var uuids []string
+		for _, devID := range req.DeviceIDs {
+			id, err := strconv.Atoi(devID)
+			if err == nil {
+				ids = append(ids, id)
+				continue
+			}
+			// if not an integer, then assume UUID.
+			uuids = append(uuids, devID)
+		}
+		if len(ids) > 0 {
+			opts = append(opts, nvidia.WithDevices(ids...))
+		}
+		if len(uuids) > 0 {
+			opts = append(opts, nvidia.WithDeviceUUIDs(uuids...))
+		}
+	}
+
+	if req.Count == "all" {
+		opts = append(opts, nvidia.WithAllDevices)
+	} else if req.Count != "" {
+		count, err := strconv.Atoi(req.Count)
+		if err != nil {
+			return err
+		}
+		opts = append(opts, nvidia.WithDevices(countToDevices(count)...))
+	}
+
+	var nvidiaCaps []nvidia.Capability
+	// req.Capabilities contains device capabilities, some but not all are NVIDIA driver capabilities.
+	for _, c := range dev.selectedCaps {
+		nvcap := nvidia.Capability(c)
+		if _, isNvidiaCap := allNvidiaCaps[nvcap]; isNvidiaCap {
+			nvidiaCaps = append(nvidiaCaps, nvcap)
+			continue
+		}
+		// TODO: nvidia.WithRequiredCUDAVersion
+		// for now we let the prestart hook verify cuda versions but errors are not pretty.
+	}
+
+	if nvidiaCaps != nil {
+		opts = append(opts, nvidia.WithCapabilities(nvidiaCaps...))
+	} else {
+		// TODO: shouldn't the default be something more restricted than AllCapabilities ?
+		opts = append(opts, nvidia.WithAllCapabilities)
```

Comment:
Can we set the default to no capabilities?

A first argument is that this prevents us from introducing breaking changes in the future if we want to change this. 
A second argument is that people using GPU images currently have the requirement to set this variable, so this is the current behavior :)